### ISSUE TITLE
Plugin can produce empty commits and committed files show up as modified in git status

### ISSUE THREAD

**Issue by pfeileon** [Cosium/git-code-format-maven-plugin#96]
**What steps will reproduce the problem?**
1. Make changes to a file's formatting which conflict with google-code-formatter.
2. Commit the changes.

**Expected**
- There should be no commit as there are no changes after the formatter fixed the files in the pre-commit hook.
- `git status` should report a clean state. 

**Actual**
- An empty commit was produced.
- `git status` reports the file as modified, `git diff` reports no changes, however.

The part with the modified files without changes showing up also happens after a "real" commit. `git add .` cleans it up again.

**Environment:**
git-code-format-maven-plugin 4.2
Apache Maven 3.8.6
Java version: 11.0.14 (Red Hat)
Windows 10
git version 2.38.0.windows.1

**Comment by mjuergens-cg** [comment]
See https://github.com/Cosium/git-code-format-maven-plugin/issues/227

**Comment by RafaelMRG** [comment]
Anyone managed to fix this issue ? i am pretty much running almost the same setup as described and get the same git behavior

**Comment by reda-alaoui** [comment]
This issue reports two distinct problems. One is fixed, the other one belongs to git itself.

**Committed files showing up as modified.** This one was the same bug as #227 and is fixed by #381,
which will be part of the next release.

**The empty commit.** This one still happens, and it cannot be solved in this plugin, only worked
around.

Git does reject a commit holding no change, but it performs that check *before* it re-reads the index
the pre-commit hook modified. From `builtin/commit.c`, in `prepare_to_commit()`: the pre-commit hook
runs first, `committable` is then computed from the in-memory index, the empty commit is rejected a
few lines below, and only after that comes:

```c
	if (!no_verify && invoked_hook) {
		/*
		 * Re-read the index as the pre-commit-commit hook was invoked
		 * and could have updated it. We must do this before we invoke
		 * the editor and after we invoke run_status above.
		 */
		discard_index(&the_index);
	}
	read_index_from(&the_index, index_file, get_git_dir());
```

So the emptiness check always sees the index as it was *before* the hook, while the tree that ends up
being committed is built from the index as the hook left it. When the formatting reverts everything
that was staged, git has no idea and records the commit. I reproduced it on git 2.43.0, with both
`git commit -m` and the editor path.

The hook cannot report the situation either. Per `githooks(5)`, the pre-commit hook "takes no
parameters" and "exiting with a non-zero status from this script causes the git commit command to
abort" — the exit status is its only channel, and there is no way to say "there is nothing left to
commit, stop here" as opposed to "this commit is rejected".

That leaves failing the commit as the only lever, and that cannot be the default behaviour: it would
bring back aborted commits, which #379 deliberately moved away from — a commit is no longer aborted
when the formatting cannot be merged into the unstaged changes, it succeeds with a warning. The hook
also has no way to know that `--allow-empty` was passed, since it receives no parameters, so it would
refuse even a commit you explicitly asked to be empty.

This is not new upstream. The very same behaviour was reported to the git mailing list in 2014,
[Using "git reset" in a pre-commit hook can create an empty commit](https://www.mail-archive.com/git@vger.kernel.org/msg44005.html),
with the expectation that "Git should refuse to continue if the pre-commit hook resets all changes".
It got no answer and nothing changed since. Giving the pre-commit hook parameters has been proposed
at least twice, in 2010 and again in 2014 with
[commit: inform pre-commit if --amend is used](https://www.mail-archive.com/git@vger.kernel.org/msg61555.html),
reviewed favourably by Jeff King and Junio C Hamano, and it never landed.

A proper fix belongs to git: either re-evaluating whether the commit is empty after the index has been
re-read, or giving the hook a way to signal it.

Keeping this issue open for that second half. The plugin cannot stop git from recording the empty
commit, but I will add an opt-in option, something like `gcf.failOnEmptyCommit` defaulting to
`false`, making the hook fail the commit when the formatting leaves nothing to commit. It will always
ignore `--allow-empty`, with `--no-verify` as the only way past it.
